[WIP] Subscribe/Unsubscribe

Huddie opened the pull request:
Closes #2273 

So I think I have everything in place. Just need to configure the request that subscribes and unsubscribes. 

Where I'm a bit stuck:

There is a difference between watching repositories (https://developer.github.com/v3/activity/watching/) vs subscribing a thread  https://developer.github.com/v3/activity/notifications/). 

I think we're looking for the latter and so I'm trying to figure out how to obtain the thread_id from the issueOrPullRequest query that we use. We have  `V3SubscribeThreadRequest` which takes an `id` and `ignore`. So once I get the thread_id I can pass it in. 

Anyone know where I can get the thread_id from the issueOrPullRequest query response?

Sherlouk commented on Classes/Issues/PullRequest+IssueType.swift:
```diff
@@ -12,6 +12,10 @@ import StyledTextKit
 extension IssueOrPullRequestQuery.Data.Repository.IssueOrPullRequest.AsPullRequest: IssueType {
 
+    var viewerSubscribed: Bool {
+        return viewerSubscription?.rawValue != "IGNORED"
```
nit: We should use a constant for this value here: `SubscriptionState.ignored.rawValue`

Sherlouk commented on Classes/Issues/Issue+IssueType.swift:
```diff
@@ -11,6 +11,10 @@ import IGListKit
 extension IssueOrPullRequestQuery.Data.Repository.IssueOrPullRequest.AsIssue: IssueType {
 
+    var viewerSubscribed: Bool {
+        return viewerSubscription?.rawValue != "IGNORED"
```
nit: We should use a constant for this value here: `SubscriptionState.ignored.rawValue`

Huddie replied:
I shall fix this!

Huddie commented:
@Sherlouk I tried that. I put a print statement to get that ID and I put a print to get the ID of the issue in the inbox and the ids were different. One was: 182635 (or something like that, all numbers) and then the other was the MDEx.... iD 🤔. Thoughts?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

Sherlouk commented:
If it's all numbers it sounds like it could just be the PR/Issue number (e.g. this one is #2482) I believe it should just be the `MDEx` one - what error are you seeing?

There don't appear to be any other IDs 🤔

Huddie commented:
In the docs you can see it’s called a thread_id rather than a regular ID, not really sure where to get that/what that is. In the NotificationModelController it’s called v3id, so seems to be some V3 API. If you capture the v3id you get something like:

415678928 (this is an actual example for this thread)


<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

Huddie commented:
Well, this is frustrating. I ended up posting on the Github Dev Forum. So close to finishing this :/

Sherlouk commented:
@Huddie I know it's a bit of a side step from what you were looking into but _could_ we maybe investigate using the V4 API mutation for this? Documented [here](https://developer.github.com/v4/mutation/updatesubscription/)

It looks like we have all the information necessary for this one?

Sherlouk commented:
@Huddie It's been a while so I don't know how much I can help specifically but if you look at the add comment flow or the add/remove reaction flow then both of these use V4 mutations to work! 

Should be able to reuse/take notes from those implementations!

If you have any specific questions though feel free to post them and I'm sure someone here will be able to help if not myself!

Huddie commented:
@Sherlouk great idea with V4 mutation. The mutation I added (UpdateSubscription.graphql) works in graphiQL

```
mutation UpdateSubscription($subscribable_Id: ID!, $subscription_state: SubscriptionState!) {
  updateSubscription(input:{subscribableId: $subscribable_Id, state: $subscription_state}) {
    subscribable{
      id
      viewerCanSubscribe
      viewerSubscription
    }
  }
}

{
   "subscribable_Id": "MDU6SXNzdWUzODU1MTAyMTU=",
   "subscription_state": "IGNORED"
}
```

but doesn't seem to work in-app. I'm not getting back an error, just nil and nothing is changing on github. When I used GraphiQL I am able to change the subscription status on github. Any ideas?

Sherlouk commented:
Hmm 😬 Interesting that it works in one but not the other - one would think it has something to do with the inputs at that point?

Maybe @rnystrom can help here, has a lot more knowledge!

Huddie commented:
@Sherlouk The inputs I put in the GraphiQL were taken from the console after running the app and printing the inputs. So they should be exactly the same. I’ll look into it more

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

Huddie commented:
@Sherlouk Got it working, dumb mistake on my part. 
@rnystrom my mutation is working on repos that I owe but when I try subscribing to an issue on GitHawk or a random repo I get 

> "Although you appear to have the correct authorization credentials, the GitHawkApp organization has enabled OAuth app access restrictions, meaning that data access to third-parties is limited"

why would subscribing to threads require any special access? If anything I feel like sub/unsub should be open to all repos regardless of access. Plus I can sub/unsub on github.com so it must be a mistake i'm making. Any ideas?

Huddie commented:
@ijm8710 The only necessary permission im pretty sure is 'notifications'

ijm8710 commented:
Yeah exactly, I would be shocked if you run into that issue with the pat

ijm8710 commented:
Yeah this all goes back to what ryan and I summed up in #2376 per what we investigated in #2165. Imo, unless GitHub changes how the oath works (because I think all of us agree the OAuth alone should grant sufficient access) the permissions should be reset for all current githawk users and reapplied going forward for all new githawk users to grant permissions on a token and not OAuth level

Huddie commented:
Well think stinks. @rnystrom Any way around this or am I putting this pr on the shelf  ?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

rnystrom commented:
I don’t know if it’s a blocker for this, the same error message occurs for other actions, like reactions, right?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

Huddie commented:
@rnystrom right. Ya I forgot that for a second. Okay so I guess I’ll keep going. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

ijm8710 commented:
It’s the same error message you get even for posting comments in some repos as well. 

I don’t understand why you’d put this on the shelf at all. Again the awesome work on this is not really being newly blocked by anything. The PAT issues have been existent for a while in things even more critical than this (commenting)

If anything that just increases priority on fixing PAT but again no issue here.

Huddie commented:
@ijm8710 Thanks!

Well... in that case @rnystrom this is ready for review 😀

General changes:
1. Manager button will *Always* show. 
    - Reason for this is that you can always subscribe/unsubscribe to an issue
2. Subscribe/Unsubscribe is the first cell under the divider in the manager menu, not locked/unlocked. 
    - Reason: Seemed to make more sense, especially since sometimes sun/unsub is the only option.
3. Added viewerSubscription to API request for Issues/PR's.
    - Reason: We need this information for determining the current subscription status 
4. Added a new mutation that allows for changing users subscription status

ijm8710 commented:
Just quick food for thought is where the edit title would rank relating to your second bullet since it seems that PR was close to completed. Honestly I think the staticness of the button raises even more support for the addition of the mark read button inta-post being innthe burger as welll. I think there could be a bunch of viable options perhaps for the burger menu now that it’s always visible👍

Huddie commented:
@ijm8710 Not sure I fully follow but does this have to do with this PR specifically or is it something that should be addressed later? Just want to make sure this thread isn’t about future idea/features and is strictly about this PR.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

Huddie commented:
@rnystrom 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

rnystrom commented:
Haven't had time to catch up. @Huddie what's the current status w/ this?

Huddie commented:
@BasThomas Can I get a review on this. Might have time next week to fix it up if there are requested changes. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

BasThomas commented on Classes/Issues/GithubClient+Issues.swift:
```diff
@@ -297,6 +298,37 @@ extension GithubClient {
+    func setSubscription(
+        previous: IssueResult,
+        subscribe: Bool,
+        completion: ((Result<SubscriptionState?>) -> Void)? = nil
```
Hm, what would it mean if the `state` is `nil`? Would that be an actual success?

BasThomas commented on Classes/Issues/GithubClient+Issues.swift:
```diff
@@ -297,6 +298,37 @@ extension GithubClient {
+        let cache = self.cache
+        cache.set(value: optimisticResult)
+
+        let mutation = UpdateSubscriptionMutation(subscribable_Id: previous.id, subscription_state: state)
```
Can we change these parameters to `camelCase`?

BasThomas commented on Classes/Issues/IssueManagingContextController.swift:
```diff
@@ -144,13 +155,19 @@ final class IssueManagingContextController: NSObject, ContextMenuDelegate {
+            iconName = "mute"
         }
 
         // Lock always has the divider above it assuming you're a collaborator.
```
We should update this comment.

BasThomas commented on FreetimeTests/SplitViewTests.swift:
```diff
@@ -114,9 +114,7 @@ class SplitViewTests: XCTestCase {
             client: GithubClient(userSession: nil),
             repo: RepositoryDetails(
                 owner: "Foo",
-                name: "Bar",
```
Why was this changed? 🤔

Huddie commented:
@BasThomas can you explain where you think we have this again? Not sure I fully understood

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

BasThomas commented:
![GitHawk Upload by BasThomas](https://i.imgur.com/vyiyS7e.jpg)

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

Huddie commented:
@BasThomas ya so that’s true for notifications. If you comment on a issue etc etc you get subscribed and then once you get the notification you can unsubscribe. Here you can pick any issue and subscribe to it, and the. Unsubscribe whenever as wel, even if you hasn’t gotten a notification. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>
